book/src/development/method_checks.md

```diff
@@ -0,0 +1,92 @@
+# Method Checks
+
+In some scenarios we might want to check for methods when developing
+a lint. There are two kinds of questions that we might be curious about:
+
+- Invocation: Does an expression call a specific method?
+- Definition: Does the type `Ty` of an expression define a method?
+
+## Checking if an expr is calling a specific method
+
+Suppose we have an `expr`, we can check whether it calls a specific
+method, e.g. `our_fancy_method`, by performing a pattern match on
+the [ExprKind] that we can access from `expr.kind`:
+
+```rust
+use rustc_hir as hir;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_span::sym;
+
+impl<'tcx> LateLintPass<'tcx> for OurFancyMethodLint {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'_>) {
+        // Check our expr is calling a method with pattern matching
+        if let hir::ExprKind::MethodCall(path, _, [_self_arg, ..]) = &expr.kind
+            // Check the name of this method is `some_method`
+            && path.ident.name == sym!(our_fancy_method)
+            // Optionally, we can check the type of the self argument whenever necessary.
+            // See "Type Checking" chapter of the Clippy book for more information.
```

Comment:
It's good that you mentioned this, since this is an easy thing to overlook and already came up a few times in my reviews. But, I think this should be formulated stronger. It's almost never correct to only check for the name. For example, just because the method is called `map` doesn't mean that it actually belongs to an `Iterator`. It could be an option or a user-defined type and method. Maybe we can directly add the type check, even if that overloads the example a bit :thinking: